vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,666 @@
+// the output_example function has special logic to ensure the number of feaures is correctly calculated.
+// Copy anything in default namespace for slot to ccb_slot_namespace in shared
+// Copy other slot namespaces to shared
+void inject_slot_features(example* shared, example* slot)
+{
+  for (auto index : slot->indices)
+  {
+    // constant namespace should be ignored, as it already exists and we don't want to double it up.
+    if (index == constant_namespace)
+    {
+      continue;
+    }
+    else if (index == default_namespace)  // slot default namespace has a special namespace in shared
+    {
+      LabelDict::add_example_namespace(*shared, ccb_slot_namespace, slot->feature_space[32]);
+    }
+    else
+    {
+      LabelDict::add_example_namespace(*shared, index, slot->feature_space[index]);
+    }
+  }
+}
+
+template <bool audit>
+void inject_slot_id(ccb& data, example* shared, int id)
+{
+  // id is zero based, so the vector must be of size id + 1
+  if(id + 1 > data.hashes.size())
+  {
+    data.hashes.resize(id + 1, 0);
+  }
+
+  uint64_t index;
+  if(data.hashes[id] == 0)
+  {
+    auto current_index_str = "index"+std::to_string(id);
+    index = VW::hash_feature(*data.all, current_index_str, data.id_namespace_hash);
+    data.hashes[id] = index;
+  }
+  else
+  {
+    index = data.hashes[id];
+  }
+
+  shared->feature_space[ccb_id_namespace].push_back(1., index);
+  shared->indices.push_back(ccb_id_namespace);
+
+  if (audit)
+  {
+    auto current_index_str = "index"+std::to_string(id);
+    shared->feature_space[ccb_id_namespace].space_names.push_back(audit_strings_ptr(new audit_strings(data.id_namespace_str, current_index_str)));
+  }
+}
+
+// Since the slot id is the only thing in this namespace, the popping the value off will work correctly.
+template<bool audit>
+void remove_slot_id(example* shared)
+{
+  shared->feature_space[ccb_id_namespace].indicies.pop();
+  shared->feature_space[ccb_id_namespace].values.pop();
+  shared->indices.pop();
+
+  if (audit)
+  {
+    shared->feature_space[ccb_id_namespace].space_names.pop();
+  }
+}
+
+void remove_slot_features(example* shared, example* slot)
+{
+  for (auto index : slot->indices)
+  {
+    // constant namespace should be ignored, as it already exists and we don't want to double it up.
+    if (index == constant_namespace)
+    {
+      continue;
+    }
+
+    else if (index == default_namespace)  // slot default namespace has a special namespace in shared
+    {
+      LabelDict::del_example_namespace(*shared, ccb_slot_namespace, slot->feature_space[32]);
+    }
+    else
+    {
+      LabelDict::del_example_namespace(*shared, index, slot->feature_space[index]);
+    }
+  }
+}
+
+// Generates quadratics between each namespace and the slot id as well as appends slot id to every existing interaction.
+void calculate_and_insert_interactions(example* shared, std::vector<example*> actions, std::vector<std::string>& vec)
+{
+  std::vector<std::string> new_interactions;
```

Comment:
Use static thread_local, with clear pattern to avoid allocation here?

For example:
[https://github.com/rajan-chari/vowpal_wabbit/blob/5c21b2dfda90cd8beb1372a0926f4a445dffb708/vowpalwabbit/offset_tree.cc#L128](https://github.com/rajan-chari/vowpal_wabbit/blob/5c21b2dfda90cd8beb1372a0926f4a445dffb708/vowpalwabbit/offset_tree.cc#L128)